### PR TITLE
Lint frontend files

### PR DESCRIPTION
**Pull request by superlopuh**
Did a pass in the frontend folder, some are straightforwad, some not so much, one remains.

### CODE REVIEW

File: xdsl/frontend/context.py [hunk 1]
```diff
@@ -3,14 +3,15 @@
...
+from sys import _getframe  #type: ignore
+from typing import Any  # type: ignore
 
 from xdsl.frontend.program import FrontendProgram
 from xdsl.frontend.python_code_check import PythonCodeCheck
 
 
 @dataclass
-class CodeContext(AbstractContextManager):
+class CodeContext(AbstractContextManager[Any]):
```
Comment (superlopuh) [inline xdsl/frontend/context.py:14]: No idea what the generic argument does, I did a quick search on GitHub and this was the first example of generics annotations on `AbstractContextManager` that I found...

File: xdsl/frontend/dialects/arith.py [hunk 2]
```diff
@@ -4,86 +4,86 @@
 from xdsl.frontend.dialects.builtin import index, i1, i32, i64, f16, f32, f64
 from xdsl.ir import Operation
 
-_Int = TypeVar("_Int", bound=Union[index, i1, i32, i64], covariant=True)
+Int = TypeVar("Int", bound=Union[index, i1, i32, i64])
```
Comment (superlopuh) [inline xdsl/frontend/dialects/arith.py]: The type is used in the interface of a public function, feels like it should be visible to clients. But now that I look at it maybe, I can change it back. Will try that out rn

File: xdsl/frontend/python_code_check.py [hunk 3]
```diff
@@ -298,6 +325,7 @@ def visit_FunctionDef(self, node: ast.FunctionDef) -> ast.FunctionDef:
 
     def visit_Name(self, node: ast.Name) -> ast.Name:
         if node.id == self.name:
+            # TODO: @George what should the type be?
```
Comment (superlopuh) [inline xdsl/frontend/python_code_check.py:328]: @georgemitenkov  ?

Comment (superlopuh) [reply xdsl/frontend/python_code_check.py:328]: Feels like this might be an actual bug, although I'm probably misunderstanding something

File: tests/test_frontend_type_conversion.py [hunk 4]
```diff
@@ -1,11 +1,11 @@
 from __future__ import annotations
 
 import ast
-from sys import _getframe
+from sys import _getframe  # type: ignore
 from typing import Any, Callable, Generic, Literal, Tuple, TypeAlias, TypeVar
 import pytest
 
-from xdsl.frontend.dialects.builtin import _FrontendType
+from xdsl.frontend.dialects.builtin import _FrontendType  # type: ignore
```
Comment (math-fehr) [inline tests/test_frontend_type_conversion.py:8]: Why do you need a type: ignore here?

File: tests/test_frontend_type_conversion.py [hunk 5]
```diff
@@ -57,6 +57,7 @@ def to_xdsl() -> Callable[..., Any]:
 
 def test_raises_exception_on_unknown_type():
     type_converter = TypeConverter(globals)
+    type_converter.file = 'file'
```
Comment (math-fehr) [inline tests/test_frontend_type_conversion.py:60]: Why is this needed?

File: xdsl/frontend/code_generation.py [hunk 6]
```diff
@@ -49,9 +49,18 @@ class CodegGenerationVisitor(ast.NodeVisitor):
     because inner functions and global variables are not allowed (yet).
     """
 
-    file: str = field(default=None)
+    _file: str | None = field(default=None)
```
Comment (math-fehr) [inline xdsl/frontend/code_generation.py:52]: Is making this optional the only solution?
I think this is supposed to never be `None`, since it is assigned in the `__init__`, right?
Could we just remove the default value instead?

File: xdsl/frontend/context.py (hunk 1 above)
Comment (math-fehr) [reply xdsl/frontend/context.py:14]: Yep, having a `Any` should always be fine!

File: xdsl/frontend/dialects/arith.py [hunk 7]
```diff
@@ -4,86 +4,86 @@
 from xdsl.frontend.dialects.builtin import index, i1, i32, i64, f16, f32, f64
 from xdsl.ir import Operation
 
-_Int = TypeVar("_Int", bound=Union[index, i1, i32, i64], covariant=True)
```
Comment (math-fehr) [inline xdsl/frontend/dialects/arith.py:7]: I would check with @georgemitenkov  if this is fine or not

File: xdsl/frontend/dialects/builtin.py [hunk 8]
```diff
@@ -15,10 +15,14 @@ def to_xdsl() -> Callable[..., Any]:
 
 
 # Type parameters for integers.
-_Width = TypeVar("_Width", bound=int, covariant=True)
```
Comment (math-fehr) [inline xdsl/frontend/dialects/builtin.py:18]: Same here

File: xdsl/frontend/python_code_check.py [hunk 9]
```diff
@@ -104,11 +113,20 @@ def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
 @dataclass
 class MultipleScopeVisitor(ast.NodeVisitor):
 
-    file: str = field(default=None)
+    _file: str | None = field(default=None)
```
Comment (math-fehr) [inline xdsl/frontend/python_code_check.py:116]: Same here, the `__init__` will always set `file`

File: xdsl/frontend/type_conversion.py [hunk 10]
```diff
@@ -39,13 +39,22 @@ class TypeConverter:
     what overloaded Python operations does this xDSL type support.
     """
 
-    file: str = field(default=None)
+    _file: str | None = field(default=None)
```
Comment (math-fehr) [inline xdsl/frontend/type_conversion.py:42]: Same here

**Review by math-fehr — COMMENTED** [review]
Nice! I added some comments

File: tests/test_frontend_type_conversion.py (hunk 4 above)
Comment (superlopuh) [reply tests/test_frontend_type_conversion.py:8]: underscore means private, it doesn't like that

File: tests/test_frontend_type_conversion.py (hunk 5 above)
Comment (superlopuh) [reply tests/test_frontend_type_conversion.py:60]: If the file is None it asserts False

File: xdsl/frontend/code_generation.py (hunk 6 above)
Comment (superlopuh) [reply xdsl/frontend/code_generation.py:52]: It's not assigned in the init, and I'm not sure exactly how it's assigned. I wanted to keep as much of the functionality the same as possible. This is potentially worth an issue to follow up, and make it non-optional

File: xdsl/frontend/dialects/arith.py (hunk 7 above)
Comment (superlopuh) [reply xdsl/frontend/dialects/arith.py:7]: Feels like we're never going to get subclasses of i1 etc, we could probably revisit if we think it might be possible. It definitely doesn't like passing covariant arguments

File: xdsl/frontend/python_code_check.py (hunk 9 above)
Comment (superlopuh) [reply xdsl/frontend/python_code_check.py:116]: - [ ] Check the files again

File: xdsl/frontend/code_generation.py (hunk 6 above)
Comment (superlopuh) [reply xdsl/frontend/code_generation.py]: Fixed

File: xdsl/frontend/type_conversion.py (hunk 10 above)
Comment (superlopuh) [reply xdsl/frontend/type_conversion.py:42]: Actually this one is not set in the init, will keep my change for now. Would be good to make it required in a different PR, but keeping current behaviour to minimise changes

File: tests/test_frontend_type_conversion.py (hunk 5 above)
Comment (math-fehr) [reply tests/test_frontend_type_conversion.py:60]: Can you pass it to the constructor directly then?

File: xdsl/frontend/dialects/arith.py (hunk 7 above)
Comment (math-fehr) [reply xdsl/frontend/dialects/arith.py:7]: Okay! Fine for me.
Here, the covariant thing is mostly useful for going from `i1 | i32` to `i1` though, it's not necessary about subclasses of these operations.

Comment (superlopuh) [reply xdsl/frontend/dialects/arith.py:7]: ah, I didn't think of that!

File: xdsl/frontend/python_code_check.py (hunk 3 above)
Comment (georgemitenkov) [reply xdsl/frontend/python_code_check.py:301]: What type are you referring to here? This just compares if the should replace the identifier we are visiting (if not, return the old node)

**Review by georgemitenkov — APPROVED** [review]

File: xdsl/frontend/type_conversion.py [hunk 11]
```diff
@@ -72,7 +72,7 @@ def _convert_name(self, type_hint: ast.Name) -> Attribute:
         # First, type can be generic, e.g. `class _Integer(Generic[_W, _S])`.
         if isinstance(type_class, _GenericAlias):
             generic_type_arguments = type_class.__args__
-            arguments_for_constructor = []
+            arguments_for_constructor = list[Any]()
```
Comment (georgemitenkov) [inline xdsl/frontend/type_conversion.py:75]: Can we not have `x: List[Any] = []`?

File: xdsl/frontend/dialects/arith.py (hunk 7 above)
Comment (georgemitenkov) [reply xdsl/frontend/dialects/arith.py:7]: Yep, I think covariance came from cases like `i1 | i32` to `i1` but I do not think this matters too much, I am fine with removing it for now

File: tests/test_frontend_type_conversion.py [hunk 12]
```diff
@@ -56,7 +56,7 @@ def to_xdsl() -> Callable[..., Any]:
 
 
 def test_raises_exception_on_unknown_type():
-    type_converter = TypeConverter(globals)
+    type_converter = TypeConverter(globals, 'file')
```
Comment (georgemitenkov) [inline tests/test_frontend_type_conversion.py:59]: Why not defaulting to None though? Otherwise error messages get inconsistent and it seems weird passing this string around in tests?

File: xdsl/frontend/type_conversion.py (hunk 11 above)
Comment (superlopuh) [reply xdsl/frontend/type_conversion.py:75]: Sure they're equivalent, I'm not sure which I prefer...

File: tests/test_frontend_type_conversion.py (hunk 12 above)
Comment (superlopuh) [reply tests/test_frontend_type_conversion.py:59]: It's just that bits of code assume it's not `None`. I can have a go at making it all optional and seeing where that goes

- [ ] make file optional and see where that breaks

File: xdsl/frontend/python_code_check.py (hunk 3 above)
Comment (superlopuh) [reply xdsl/frontend/python_code_check.py:301]: It might be easier if you check out the branch and run the pyright locally. The types of the python objects don't match

File: xdsl/frontend/type_conversion.py (hunk 11 above)
Comment (superlopuh) [reply xdsl/frontend/type_conversion.py]: Changed it to the other way

File: tests/test_frontend_type_conversion.py (hunk 12 above)
Comment (superlopuh) [reply tests/test_frontend_type_conversion.py]: Yeah it's also ok being optional everywhere

File: xdsl/frontend/python_code_check.py [hunk 13]
```diff
@@ -298,6 +298,7 @@ def visit_FunctionDef(self, node: ast.FunctionDef) -> ast.FunctionDef:
 
     def visit_Name(self, node: ast.Name) -> ast.Name:
```
Comment (georgemitenkov) [inline xdsl/frontend/python_code_check.py:299]:
```suggestion
    def visit_Name(self, node: ast.Name) -> ast.Name | ast.Constant:
```

File: xdsl/frontend/python_code_check.py (hunk 3 above)
Comment (georgemitenkov) [reply xdsl/frontend/python_code_check.py:301]: So, this visitor returns a new AST node which replaces the old one, and we want to replace the name with evaluated constant. I added a fix above.

**Review by georgemitenkov — APPROVED** [review]

**Review by math-fehr — APPROVED** [review]
Very nice! Thanks George for reviewing it!
